tests/test_streaming.py

```diff
@@ -12,6 +12,103 @@
+def test_dataloader_epoch_size_no_streams(local_remote_dir: Any, batch_size: int, seed: int,
+                                          shuffle: bool, drop_last: bool, num_canonical_nodes: int,
+                                          epoch_size: int):
+    remote_dir, local_dir = local_remote_dir
+    convert_to_mds(out_root=remote_dir,
+                   dataset_name='sequencedataset',
+                   num_samples=117,
+                   size_limit=1 << 8)
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(local=local_dir,
+                               remote=remote_dir,
+                               shuffle=shuffle,
+                               batch_size=batch_size,
+                               shuffle_seed=seed,
+                               num_canonical_nodes=num_canonical_nodes,
+                               epoch_size=epoch_size)
+
+    # Build DataLoader
+    dataloader = StreamingDataLoader(dataset=dataset,
+                                     batch_size=batch_size,
+                                     num_workers=0,
+                                     drop_last=drop_last)
+
+    samples_seen = 0
+    for batch in dataloader:
+        samples_seen += batch['sample'].size(dim=0)
+
+    if drop_last:
+        assert samples_seen == epoch_size - (epoch_size % batch_size)
+    else:
+        assert samples_seen == epoch_size
+
+
+@pytest.mark.parametrize('batch_size', [1, 4])
+@pytest.mark.parametrize('seed', [2222])
+@pytest.mark.parametrize('shuffle', [False, True])
+@pytest.mark.parametrize('drop_last', [False])
+@pytest.mark.parametrize('num_canonical_nodes', [1])
+@pytest.mark.parametrize('epoch_size', [10, 100])
+@pytest.mark.usefixtures('local_remote_dir')
+def test_dataloader_fixed_sampling(local_remote_dir: Any, batch_size: int, seed: int,
+                                   shuffle: bool, drop_last: bool, num_canonical_nodes: int,
+                                   epoch_size: int):
+    remote_dir, local_dir = local_remote_dir
+    convert_to_mds(out_root=remote_dir,
+                   dataset_name='sequencedataset',
+                   num_samples=117,
+                   size_limit=1 << 8)
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(local=local_dir,
+                               remote=remote_dir,
+                               shuffle=shuffle,
+                               batch_size=batch_size,
+                               shuffle_seed=seed,
+                               num_canonical_nodes=num_canonical_nodes,
+                               epoch_size=epoch_size,
+                               sampling_method='fixed')
+
+    # Build DataLoader
+    dataloader = StreamingDataLoader(dataset=dataset,
+                                     batch_size=batch_size,
+                                     num_workers=0,
+                                     drop_last=drop_last)
+
+    # iterate once over the dataloader (first epoch)
```

Comment:
You can simplify the two for loop into one something like this: I haven't ran it, maybe you need to take care of shallow copy / deep copy !!

```
# iterate once over the dataloader (first epoch)
first_samples_seen = {}
second_samples_seen = {}
is_first_epoch = True
for batch in dataloader:
    samples_seen = first_samples_seen if is_first_epoch else second_samples_seen
    for element in batch['sample']:
        int_element = int(element)
        if int_element in samples_seen:
            samples_seen[int_element] += 1
        else:
            samples_seen[int_element] = 1
    samples_seen = {}
```

Reply:
The dataloader only has enough batches for a single epoch I thought? so you would have to loop twice right? To my understanding the dataloader restarts at the beginning of the epoch when iterated over again. I'll definitely remove the redundancy in the for loops though. The elements are the just ints and dictionary equality should work...